### ISSUE TITLE
[Bug]: DeepSeek-V4.1 produces NaN with CUDA graphs on SM120/121 — dummy capture batches poison the null KV block

### ISSUE THREAD

**Issue by sumsliu** [vllm-project/vllm#57156]
## Your current environment

```
vLLM            main @ b443c1c (built from source)
FlashInfer      0.6.18.post1
PyTorch         2.13 + CUDA 13.0
GPU             8x NVIDIA GB10 (sm_121), 121.6 GiB unified memory each
OS              Ubuntu 24.04 aarch64
Model           DeepSeek-V4.1-Flash, TP8 + EP8
Backend         FLASHINFER_MLA_SPARSE_DSV41
KV cache        fp8_ds_mla, indexer MXFP4, block size 64 (SWA cache 32)
```

## Describe the bug

With CUDA graphs enabled, `DeepseekV41ForCausalLM` returns empty or garbled completions.
`--enforce-eager` is correct. All graph modes fail identically: `FULL`, `PIECEWISE`, and
`FULL_DECODE_ONLY`.

Eager, correct:

```
prompt: "The capital of France is"   ->  " Paris."
```

PIECEWISE, same build and weights:

```
prompt: "The capital of France is"   ->  ""
```

## Root cause

**Every attention output row is NaN from the first decoder layer onward, while its inputs are
finite.** A replay-time probe on the breakable-CUDA-graph segments shows layer 0 receiving
finite `hidden_states`, `q` and `kv`, then writing NaN into the real-token rows of its output
buffer.

The NaN comes from the KV cache, specifically block 0.

`sparse_mla_sm120/decode_dsv4_kernel.cuh` deliberately gathers slot 0 for invalid indices and
relies on softmax masking to discard it. The scale load is guarded:

```cpp
// line 272
if (idx_raw >= 0) {
  ...
  scale_word = __ldg(...);
}
```

but the NoPE/RoPE data load is not — it clamps instead:

```cpp
// line 299
const int idx = (idx_raw >= 0) ? idx_raw : 0;
const int block_idx_g = idx / section_pbs;
```

The intent is documented at line 431:

> Invalid = position past section_len OR slot id = -1 (indexer-padded; IO already gathered
> slot 0 into smem with idx clamped — masking to -inf kills it in softmax).

Masking sets the logit to `-1e30`, so softmax yields `p ≈ 0`. But the PV accumulation still
computes `p * v`, and in IEEE arithmetic `0 * NaN = NaN`. In MLA the NoPE tensor serves as both
K and V, so non-finite data in slot 0 propagates into `acc_nope` regardless of the mask.

**The kernel therefore has an undocumented precondition: block 0 must hold finite data.**

Measured contents of block 0, same build, immediately before the first real request:

| Mode | slot 0 NoPE | slot 0 RoPE | scales |
|---|---|---|---|
| eager | 0.0 | 0.0 | all zero |
| PIECEWISE | **NaN** | **~3.2e35** | all zero |

CUDA graph capture is what pollutes it. Dummy capture batches carry an all-zero block table,
so writes land in block 0 while the graph is being recorded.

## Reproduction of the causal step

Rather than argue from correlation, I copied the cache and re-ran the same kernel with the same
inputs, changing only block 0:

| Variant | Result |
|---|---|
| E: block 0 as-is | output rows 0-4 NaN |
| F: block 0 zeroed | output finite and correct |

Cloning every input tensor (`q`, `sparse_indices`, `swa_topk_lens`, `sinks`, `out`,
`workspace_buffer`) and re-running reproduced the NaN, which rules out the inputs.

Also ruled out, each tested individually: stream mismatch (`torch.cuda.synchronize()` and
`tvm_ffi.use_torch_stream()` both made no difference), the autotuner's `chunks_per_block`
tactic (tested -1, 1, 2, 3, 8), page-block-size handling, and widening the eager region
(`VLLM_V41_WIDE_EAGER=1`).

## Workaround

Zeroing block 0 before each kernel call fixes it completely. Block 0 is vLLM's reserved
`null_block` (`vllm/v1/core/block_pool.py`, popped at init and never handed to a request), so
this is safe:

```python
for key in ("swa_kv_cache", "compressed_kv_cache"):
    cache = kw.get(key)
    if cache is not None:
        cache[0].zero_()
```

Cost is about 18 KB of writes per layer. Single-stream throughput went from 17.2 tok/s
(`--enforce-eager`) to 32.6 tok/s with graphs enabled, and all outputs are correct.

## Open question for maintainers

I verified that block 0 holds non-finite data after capture and that zeroing it fixes the
output, but I did not trace which write produces it. `get_dummy_slot_mappings` fills with
`PAD_SLOT_ID`, and the fused SWA insert kernel does skip `slot_id < 0`, so the polluting write
is somewhere else — possibly an indexer or compressor cache insert that does not guard negative
slots. A maintainer who knows those paths could pin it down quickly.

## Suggested fix

Three options, in what I think is decreasing order of merit:

1. **Zero the null block once after graph capture completes.** Real requests never touch block 0,
   so this holds for the life of the process and costs nothing per step.
2. **Guard the data load in FlashInfer**, matching what the scale load at line 272 already does.
   Correct at the source, but adds a branch to the hot path.
3. **Prevent dummy batches from writing to block 0** during capture. Cleanest conceptually,
   but requires finding the unguarded write first.

I am happy to open a PR for whichever direction you prefer.

## Related

Not a duplicate of any of these, but they share territory and may help whoever picks this up:

- **#55636** — DeepSeek-V4 sparse MLA indexing a nonexistent block-table row during CUDA-graph
  warmup, also on DGX Spark / GB10. Same hardware and model family, and also a
  warmup-metadata problem, but it surfaces as an illegal address or a wrong KV slot in the
  Triton top-k mapper rather than NaN out of the FlashInfer decode kernel. Possibly worth
  checking whether both trace back to the same warmup path.
- **#57094** (fixes #57017) — NaN from CUDA-graph padding in Triton MLA decode. Same family,
  different mechanism: there `seq_len=0` makes stage 2 compute `0/0`; here a clamped index
  reads poisoned data and the PV accumulation computes `0 * NaN`. Both are padding slots
  interacting badly with MLA decode, which suggests the pattern is worth a general look.
- **#57028** — SM120 FlashInfer serving blockers for DSv4.1 (page geometry, multi-stream
  capture fallback). Adjacent to this code path but does not address the null-block NaN;
  I confirmed the NaN reproduces independently of page-geometry handling.

## Additional context

This surfaced while bringing DeepSeek-V4.1-Flash up on 8x DGX Spark. Full write-up, the probe
patches used to localize it, and the benchmark harness are at
https://github.com/sumsliu/dgx-spark-deepseek-v41 — see `docs/findings.md`.

**Comment by hclsys** [comment]
Can't help with the 8-way DSv4.1 repro (single GB10 here), but I went through the source on your open question and there are three things worth knowing before anyone starts hunting the polluting write.

**The null block is already being zeroed — just on the wrong side of capture.** `KVBlockZeroer.warmup()` (`vllm/v1/worker/utils.py:252-255`) does exactly `zero_block_ids([0])`. It's reached from `warmup_kernels()`, and `gpu_worker.py` calls that at line 817, five lines before `capture_model()` at 822. So block 0 gets cleaned once, *then* capture runs and dirties it, and nothing touches it again. That matches your table exactly — eager never captures, so slot 0 stays at the zeros it was left in.

That makes your option 1 a re-call of a mechanism that already exists with the right geometry (it handles the packed cross-layer DSV4 layout, `tests/v1/worker/test_dsv4_packed_zeroer_geometry.py`), not new code.

**It is a write, not uninitialised memory.** The shared backing allocation is `torch.zeros`, not `empty` (`utils.py:421`), so `~3.2e35` in the RoPE half didn't come from a dirty pool page.

**`indexer_k_quant_and_cache` isn't the culprit.** It returns on `slot_idx < 0` before any store (`csrc/libtorch_stable/cache_kernels.cu:640`), same as the SWA insert you already checked. Whatever writes block 0 is elsewhere.

On the FlashInfer side — worth noting the asymmetry you found isn't only a missing branch. Line 272 guards the scale load and line 299 clamps the data load, but in MLA the NoPE tensor is both K and V, so `0 * NaN` in the PV accumulate defeats the `-1e30` mask no matter what the index does. A guard there fixes it, but the kernel's real precondition is "slot 0 is finite", and that's worth writing down wherever the clamp stays.
